File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/SporeCloudPower.ts
import { ApplyPowerAction } from "../Action/Common/ApplyPowerAction";
import { SoundMaster } from "../Audio/SoundMaster";
import DungeonManager from "../DungeonManager";
import { OnDeathMessage } from "../Events/OnDeathMessage";
import FileHelper from "../FileHelper";
import { LocalizationProperty } from "../Gen/DB/Localization";
import { StringHelper } from "../StringHelper";
import { AbstractPower, PowerType } from "./AbstractPower";
import { VulnerablePower } from "./VulnerablePower";

export class SporeCloudPower extends AbstractPower{
    Icon: string = FileHelper.FormatPath("Powers/sporeCloud.png");
    Id: string = "Spore Cloud";
    Name: string = LocalizationProperty.Read("能力-孢子云");

    get Type(): PowerType {
        return PowerType.Buff;
    }

    public constructor(private amt:number) {
        super();
    }

    GetDescription(...args: any[]): string {
        return StringHelper.FormatColorString(LocalizationProperty.Read("能力-孢子云描述")).format(this.Amount);
    }

    protected OnDeath(msg: OnDeathMessage) {
        super.OnDeath(msg);
        if(DungeonManager.Inst.CurrentRoom.Monsters.AreMonstersBasicallyDead())
            return;
        SoundMaster.PlayVoice("SPORE_CLOUD_RELEASE");
        this.Flash(true);
        this.AddToTop(new ApplyPowerAction(msg.source,this.Owner,new VulnerablePower(),this.amt));
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Power/Buff/SporeCloudPower.cs
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using ModTheSpire.Power.Debuff;
using TCG.Config;

namespace ModTheSpire.Power.Buff
{
    public class SporeCloudPower : AbstractPower
    {
        public override string Id { get; } = nameof(SporeCloudPower);
        public override string Name { get; } = LocalizationProperty.Read("能力-孢子云");
        public override string Icon { get; } = "Powers/sporeCloud.png";
        public override PowerType Type { get; } = PowerType.Buff;
        public override string GetDescription(object[] args)
        {
            return string.Format(LocalizationProperty.Read("能力-孢子云描述"), this.Amount);
        }

        protected override void OnDeath(OnDeathMessage msg)
        {
            base.OnDeath(msg);
            if(DungeonManager.Inst.CurrentRoom.Monsters.AreMonstersBasicallyDead())
                return;
            SoundMaster.PlayVoice(SoundMaster.SPORE_CLOUD_RELEASE);
            this.Flash(true);
            this.AddToTop(new ApplyPowerAction(msg.Unit, this.Owner, new VulnerablePower(), this.Amount));
        }
    }
}